plonky2/src/batch_fri/oracle.rs

```diff
@@ -0,0 +1,610 @@
+    pub degree_bits: Vec<usize>,
+    pub rate_bits: usize,
+    pub blinding: bool,
+}
+
+impl<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>
+    BatchFriOracle<F, C, D>
+{
+    /// Creates a list polynomial commitment for the polynomials interpolating the values in `values`.
+    pub fn from_values(
+        values: Vec<PolynomialValues<F>>,
+        rate_bits: usize,
+        blinding: bool,
+        cap_height: usize,
+        timing: &mut TimingTree,
+        fft_root_table: &[Option<&FftRootTable<F>>],
+    ) -> Self {
+        let coeffs = timed!(
+            timing,
+            "IFFT",
+            values.into_par_iter().map(|v| v.ifft()).collect::<Vec<_>>()
+        );
+
+        Self::from_coeffs(
+            coeffs,
+            rate_bits,
+            blinding,
+            cap_height,
+            timing,
+            fft_root_table,
+        )
+    }
+
+    /// Creates a list polynomial commitment for the polynomials `polynomials`.
+    pub fn from_coeffs(
+        polynomials: Vec<PolynomialCoeffs<F>>,
+        rate_bits: usize,
+        blinding: bool,
+        cap_height: usize,
+        timing: &mut TimingTree,
+        fft_root_table: &[Option<&FftRootTable<F>>],
+    ) -> Self {
+        let mut degree_bits = polynomials
+            .iter()
+            .map(|p| log2_strict(p.len()))
+            .collect_vec();
+        assert!(degree_bits.windows(2).all(|pair| { pair[0] >= pair[1] }));
+
+        let num_polynomials = polynomials.len();
+        let mut group_start = 0;
+        let mut leaves = Vec::new();
+
+        for (i, d) in degree_bits.iter().enumerate() {
+            if i == num_polynomials - 1 || *d > degree_bits[i + 1] {
+                let lde_values = timed!(
+                    timing,
+                    "FFT + blinding",
+                    PolynomialBatch::<F, C, D>::lde_values(
+                        &polynomials[group_start..i + 1],
+                        rate_bits,
+                        blinding,
+                        fft_root_table[i]
+                    )
+                );
+
+                let mut leaf_group = timed!(timing, "transpose LDEs", transpose(&lde_values));
+                reverse_index_bits_in_place(&mut leaf_group);
+                leaves.push(leaf_group);
+
+                group_start = i + 1;
+            }
+        }
+
+        let batch_merkle_tree = timed!(
+            timing,
+            "build Field Merkle tree",
+            BatchMerkleTree::new(leaves, cap_height)
+        );
+
+        degree_bits.sort_unstable();
+        degree_bits.dedup();
+        degree_bits.reverse();
+        assert_eq!(batch_merkle_tree.leaves.len(), degree_bits.len());
+        Self {
+            polynomials,
+            batch_merkle_tree,
+            degree_bits,
+            rate_bits,
+            blinding,
+        }
+    }
+
+    /// Produces a batch opening proof.
+    pub fn prove_openings(
```

Comment:
If I understand correctly the logic:

You provide several `FriInstanceInfo`, which consist in multiple `FriBatchInfo`, containing each an evaluation point and a batch of polynomials to be evaluated at that point. In the batched version, for a single `FriInstanceInfo`, the evaluated polynomials can be of different degrees, contrary to the normal version.
But I don't think the inner `for` loop (at line 154) works well with polynomials of different degrees. To my understanding, what we should be doing is to group all polynomials of the same degree _across all instances_, and then compose them together with `alpha`.

Reply:
In the batched version, for a single FriInstanceInfo, the evaluated polynomials have the same degree.

Reply:
In zkVM code, we first create a FriInstanceInfo for each table. We then merge FriInstanceInfo instances when two tables have the same degree. Lastly, we have a vec<FriInstanceInfo> with different degrees, but within each FriInstanceInfo, the polynomials have the same degree.